base/src/com/google/idea/blaze/base/buildview/BazelService.kt

```diff
@@ -0,0 +1,173 @@
+      .withParameters(cmd.toArgumentList())
+      .apply { setWorkDirectory(root.pathString) } // required for backwards compatability
+      .withRedirectErrorStream(true)
+      .let(::OSProcessHandler)
+
+    handler.addProcessListener(object : ProcessListener {
+      override fun onTextAvailable(event: ProcessEvent, outputType: Key<*>) {
+        ctx.println(event.text.trimEnd())
+      }
+    })
+    handler.startNotify()
+
+    try {
+      while (!handler.isProcessTerminated) delay(100)
+    } finally {
+      handler.destroyProcess()
+    }
+
+    val exitCode = handler.exitCode ?: 1
+    if (exitCode != 0) {
+      ctx.setHasError()
+    }
+
+    return exitCode
+  }
+
+  private suspend fun parseEvent(ctx: BlazeContext, stream: BufferedInputStream) {
+    // make sure that there are at least four bytes already available
+    while (stream.available() < 4) delay(10)
+
+    // protobuf messages are delimited by size (encoded as varint32),
+    // read size manually to ensure the entire message is already available
+    val size = CodedInputStream.readRawVarint32(stream.read(), stream)
+    while (stream.available() < size) delay(10)
+
+    val eventStream = LimitedInputStream(stream, size)
+    val event = try {
+      BuildEvent.parseFrom(eventStream)
+    } catch (e: Exception) {
+      LOG.error("could not parse event", e)
+
+      // if the message could not be parsed, make sure to skip it
+      if (eventStream.bytesRead < size) {
+        stream.skip(size.toLong() - eventStream.bytesRead)
+      }
+
+      return
+    }
+
+    if (event == null) {
+      delay(10)
+    } else {
+      BuildEventParser.parse(event)?.let(ctx::output)
+    }
+  }
+
+  private fun CoroutineScope.parseEvents(ctx: BlazeContext, helper: BuildResultHelper): Job {
+    return launch(CoroutineName("EventParser")) {
+      try {
+        // wait for bazel to create the output file
+        while (!helper.outputFile.exists()) delay(10)
+
+        FileInputStream(helper.outputFile).buffered().use { stream ->
+          // keep reading events while the coroutine is active, i.e. bazel is still running,
+          // or while the stream has data available (to ensure that all events are processed)
+          while (isActive || stream.available() > 0) {
+            parseEvent(ctx, stream)
+          }
+        }
+      }
+      catch (e: CancellationException) {
+        throw e
+      }
+      catch (e: Exception) {
+        LOG.error("error in event parser", e)
+      }
+    }
+  }
+
+  fun build(ctx: BlazeContext, cmdBuilder: BlazeCommand.Builder): BlazeBuildOutputs {
+    assertNonBlocking()
+    LOG.assertTrue(cmdBuilder.name == BlazeCommandName.BUILD)
+
+    return executionScope(ctx) { provider ->
+      cmdBuilder.addBlazeFlags(provider.getBuildFlags())
+
+      val parseJob = parseEvents(ctx, provider)
+
+      val exitCode = execute(ctx, cmdBuilder.build())
+      val result = BuildResult.fromExitCode(exitCode)
+
+      parseJob.cancelAndJoin()
```

Comment:
As far as I understand this can only trigger a `CancellationException` in `parseEvent` if `parseEvent` is waiting for more data. And in this case we actually want to cancel `parseEvent` because there is not going to be more data after the bazel process finished. The actual parsing cannot be cancelled as far as I can tell.